### ISSUE TITLE
F821 – false negative for hiku

### ISSUE THREAD

**Issue by hyzyla** [charliermarsh/ruff#1690]
I'm using [hiku](https://github.com/evo-company/hiku) package for building GraphQL server, and that package use class bracket syntax for defining input parameters. 

Here is minimal example of code that I'm trying to check:

main.py
```python
from typing import Optional

from hiku.expr.core import define
from hiku.types import Any as HikuAny
from hiku.types import Optional as HikuOptional
from hiku.types import Record


@define(HikuOptional[Record[{'name': HikuAny}]])
def foo() -> Optional[str]:
    pass
``` 
requirements.txt:
```txt
hiku==0.6.0
ruff==0.0.212
```

And when I'm trying to check that code by `ruff`, I receive error:

```txt
main.py:9:30: F821 Undefined name `name`
```

**Comment by charliermarsh** [comment]
Thank you! This is revealing a bug in our import tracking (related to having both `from typing import Optional` and `from hiku.types import Optional as HikuOptional` in the same file -- it thinks that `HikuOptional` is an alias for the `Optional` from `typing`, which is of course a bug).

**Comment by hyzyla** [comment]
Here is a link [1] that point to Optional definition  in `hiku` library 

1. https://github.com/evo-company/hiku/blob/1c45037b4c2434eb73824d9506f2df9eabecf223/hiku/types.py#L119

**Comment by charliermarsh** [comment]
Thanks. Gotta think on this! Requires some refactoring, won't go out tonight but should be fixed in the next few days.

**Comment by charliermarsh** [comment]
Working on this now. It's going to be a significant refactor of how we track imports.
